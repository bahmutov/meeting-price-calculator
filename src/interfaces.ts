import xs from 'xstream';
import { VNode, DOMSource } from '@cycle/dom';
import { HTTPSource, RequestOptions } from '@cycle/http';
import { TimeSource } from '@cycle/time';
import { StorageSource, StorageRequest } from '@cycle/storage';
import { StateSource } from 'cycle-onionify';
import { State } from './components/app';

export type Sources = {
  DOM: DOMSource;
  onion: StateSource<State>;
  HTTP: HTTPSource;
  Time: TimeSource;
  storage: StorageSource;
  props$: any;
};

export type Sinks = {
  DOM: xs<VNode>;
  onion: xs<(s: any) => any>;
};

export type Component = (s: Sources) => Sinks;
