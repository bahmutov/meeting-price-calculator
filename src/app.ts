import xs, { Stream } from 'xstream';
import { VNode, div, span, img, a, h1, button } from '@cycle/dom';

import { Sources, Sinks } from './interfaces';
import Header from './components/header';
import Price from './components/price';
import { Adjustments } from './components/adjustments';
import Footer from './components/footer';

export function App(sources : Sources) : Sinks {
    const headerVDom$ : Stream<VNode> = Header(sources);
    const priceVDom$ : Stream<VNode> = Price(sources);
    const adjustmentsVDom$ : Stream<VNode> = Adjustments(sources);
    const footerVDom$ : Stream<VNode> = Footer(sources);

    const vdom$ : Stream<VNode> = xs.combine(headerVDom$, priceVDom$, adjustmentsVDom$, footerVDom$)
        .map(([headerVDom, priceVDom, adjustmentsVDom, footerVDom]) =>
            div('.App-flex-container', [
                headerVDom,
                priceVDom,
                adjustmentsVDom,
                footerVDom,
            ])
        );

    const sinks : Sinks = {
        DOM : vdom$
    };

    return sinks;
}
