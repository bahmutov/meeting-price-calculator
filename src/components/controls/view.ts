import xs from 'xstream';
import * as moment from 'moment';
import { VNode, div, select, option, span, label } from '@cycle/dom';

import { State } from './index';
import { styles } from './styles';
import { formatPrice } from '../../utils/priceUtils';

function renderOption(currency: string, selectedCurrency: string) {
  return currency === selectedCurrency
    ? option(
        {
          attrs: {
            selected: 'selected'
          }
        },
        [currency]
      )
    : option([currency]);
}

function renderCurrencySelect(selectedCurrency: string) {
  return select(`.${styles.currencySelect}`, [
    renderOption('â‚¬', selectedCurrency),
    renderOption('$', selectedCurrency)
  ]);
}

export default function view(
  state$: xs<State>,
  personAmountSliderVDom$: xs<VNode>,
  avgPriceSliderVDom$: xs<VNode>
): xs<VNode> {
  return xs
    .combine(state$, personAmountSliderVDom$, avgPriceSliderVDom$)
    .map(
      ([
        { currency, personAmount, avgPrice },
        personAmountVDom,
        avgPriceVDom
      ]) =>
        div(`.${styles.controls}`, [
          personAmountVDom,
          div(`.${styles.priceResult}`, [
            label('Total price per hour'),
            div(
              `.${styles.totalPriceValue}`,
              formatPrice(personAmount * avgPrice, currency)
            ),
            div(`.${styles.currency}`, [
              span('Currency'),
              renderCurrencySelect(currency)
            ])
          ]),
          avgPriceVDom
        ])
    );
}
