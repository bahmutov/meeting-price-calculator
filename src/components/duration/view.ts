import xs from 'xstream';
import * as moment from 'moment';

import { VNode, div, span, button } from '@cycle/dom';
import { State } from './index';
import { styles } from './styles';

export default function view(state$: xs<State>): xs<VNode> {
  return state$.map(({ startTime, duration }) =>
    div([
      div([
        span(`Start time: ${moment(startTime).format('HH:mm:ss')}`),
        button(`.${styles.resetButton}`, 'Reset')
      ]),
      div(`Duration: ${moment.duration(duration, 'seconds').humanize()}`)
    ])
  );
}
