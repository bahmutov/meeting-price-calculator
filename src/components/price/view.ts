import xs from 'xstream';
import { VNode, div, select, option, span, label } from '@cycle/dom';
import { State } from '../../interfaces';
import calculatePrice, { formatPrice } from '../../utils/priceUtils';

function renderPrice(
  personAmountValue: number,
  avgPriceValue: number,
  tick: number,
  currency: string
): VNode {
  const price: number = calculatePrice(personAmountValue, avgPriceValue, tick);
  return div('.Price-actual', [
    div('.Price-label', 'This meeting has cost'),
    div('.Price-value', formatPrice(price, currency))
  ]);
}

export default function view(
  state$: xs<State>,
  personAmountSliderVDom$: xs<VNode>,
  avgPriceSliderVDom$: xs<VNode>
): xs<VNode> {
  return xs
    .combine(state$, personAmountSliderVDom$, avgPriceSliderVDom$)
    .map(
      (
        [
          { tick, personAmount, avgPrice, currency },
          personAmountVDom,
          avgPriceVDom
        ]
      ) =>
        div('.Price', [
          renderPrice(personAmount.value, avgPrice.value, tick, currency),
          div('.PriceInputs', [
            personAmountVDom,
            div('.price-result', [
              label('.total-price-label', 'Total price per hour'),
              div(
                '.total-price-value',
                formatPrice(personAmount.value * avgPrice.value, currency)
              ),
              div('.currency', [
                span('.currency-label.label', 'Currency'),
                select('.currency-select', [option('â‚¬'), option('$')])
              ])
            ]),
            avgPriceVDom
          ])
        ])
    );
}
