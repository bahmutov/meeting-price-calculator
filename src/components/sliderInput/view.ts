import xs from 'xstream';
import { VNode, div, input, span, label } from '@cycle/dom';
import { State } from './index';
import { styles } from './styles';

export default function view(state$: xs<State>): xs<VNode> {
  return state$.map(({ description, unit, min, max, step, value }) =>
    div(`.${styles.sliderInput}`, [
      label(description),
      input(`.SliderInput-input .${styles.numberInput}`, {
        attrs: {
          type: 'number',
          min,
          max,
          step
        },
        props: { value }
      }),
      span(`.${styles.sliderInputUnit}`, unit),
      input('.SliderInput-input', {
        attrs: {
          type: 'range',
          min,
          max,
          step
        },
        props: { value }
      })
    ])
  );
}
