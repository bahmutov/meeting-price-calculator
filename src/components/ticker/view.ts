import xs from 'xstream';
import { VNode, div } from '@cycle/dom';
import { State } from './index';
import { styles } from './styles';
import { formatPrice } from '../../utils/priceUtils';

export default function view(state$: xs<State>): xs<VNode> {
  return state$.map(({ currency, totalPrice }) =>
    div(`.${styles.actualPrice}`, [
      div(`.${styles.priceLabel}`, 'This meeting has cost'),
      div(`.${styles.priceValue}`, formatPrice(totalPrice, currency))
    ])
  );
}
