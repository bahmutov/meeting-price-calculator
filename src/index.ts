import 'babel-polyfill'; //This will be replaced based on your babel-env config

import { run } from '@cycle/run';
import { makeDOMDriver } from '@cycle/dom';
import { timeDriver } from '@cycle/time';
import onionify from 'cycle-onionify';
import storageify from 'cycle-storageify';
import storageDriver from '@cycle/storage';

import { Component } from './interfaces';
import App from './components/app';

const main: Component = App;

const wrappedMain = onionify(storageify(main, { key: 'mpc-state' }));

const drivers: any = {
  DOM: makeDOMDriver('#app'),
  Time: timeDriver,
  storage: storageDriver
};

run(wrappedMain, drivers);
