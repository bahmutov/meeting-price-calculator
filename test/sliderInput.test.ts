import xs from 'xstream';
import { forall, integer, assert, Options } from 'jsverify';

import { SliderInputState } from '../src/interfaces';
import calculatePrice, { formatPrice } from '../src/utils/priceUtils';
import { VNode, div, p } from '@cycle/dom';
import { withTime } from 'cyclejs-test-helpers';
import view from '../src/components/sliderInput/view';
const htmlLooksLike = require('html-looks-like');
const toHtml = require('snabbdom-to-html');

const testOptions: Options = {
  tests: 100,
  quiet: false,
  size: 60 * 60 * 24
};

describe('SliderInput Component', () => {
  it('should match a snapshot correctly', () => {
    const state$: xs<SliderInputState> = xs.of({
      description: 'Person amount',
      unit: 'persons',
      min: 0,
      max: 100,
      step: 1,
      value: 8
    });
    const vdom$: xs<VNode> = view(state$);
    const html$: xs<{}> = vdom$.map(toHtml);
    expect(html$).toMatchSnapshot();
  });
});
