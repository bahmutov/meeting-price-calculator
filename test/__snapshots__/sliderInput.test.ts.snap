// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SliderInput Component should match a snapshot correctly 1`] = `
Stream {
  "_d": false,
  "_dl": Object {},
  "_err": Object {},
  "_ils": Array [],
  "_prod": MapOp {
    "f": [Function],
    "ins": Stream {
      "_d": false,
      "_dl": Object {},
      "_err": Object {},
      "_ils": Array [],
      "_prod": MapOp {
        "f": [Function],
        "ins": Stream {
          "_d": false,
          "_dl": Object {},
          "_err": Object {},
          "_ils": Array [],
          "_prod": FromArray {
            "a": Array [
              Object {
                "description": "Person amount",
                "max": 100,
                "min": 0,
                "step": 1,
                "unit": "persons",
                "value": 8,
              },
            ],
            "type": "fromArray",
          },
          "_stopID": Object {},
          "_target": Object {},
        },
        "out": Object {},
        "type": "map",
      },
      "_stopID": Object {},
      "_target": Object {},
    },
    "out": Object {},
    "type": "map",
  },
  "_stopID": Object {},
  "_target": Object {},
}
`;
